### PR TITLE
Features

### PR DESCRIPTION
**Pull request by AkashBoora**

### CODE REVIEW

File: java/src/main/java/com/codurance/training/tasks/TaskList.java [hunk 1]
```diff
@@ -50,17 +58,27 @@ private void execute(String commandLine) {
...
+            case "deadline":
+                String[] taskDeadline = commandRest[1].split(" ", 2);
+                taskService.addDeadlineToTask(taskDeadline[0], taskDeadline[1]);
+                break;
+            case "today":
+                showService.showDueTodayTasks(tasks);
+                break;
+            case "view":
+                view(commandRest[1]);
                 break;
```
Comment (AbhilashZemoso) [inline java/src/main/java/com/codurance/training/tasks/TaskList.java]: Separate execute method to different class

File: java/src/main/java/com/codurance/training/tasks/TaskList.java [hunk 2]
```diff
@@ -16,7 +15,11 @@ public final class TaskList implements Runnable {
     private final BufferedReader in;
     private final PrintWriter out;
 
-    private long lastId = 0;
+    ShowService showService;
+    CheckService checkService;
+    ProjectService projectService;
+    TaskService taskService;
+    DeleteService deleteService;
```
Comment (AbhilashZemoso) [inline java/src/main/java/com/codurance/training/tasks/TaskList.java]: have these private final

File: java/src/main/java/com/codurance/training/tasks/service/CheckServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+import java.io.PrintWriter;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+public class CheckServiceImpl implements CheckService {
+
+	private final PrintWriter out;
+	private Map<String, List<Task>> tasks = new LinkedHashMap<>();
```
Comment (AbhilashZemoso) [inline java/src/main/java/com/codurance/training/tasks/service/CheckServiceImpl.java]: make this final and remove initialisation

File: java/src/main/java/com/codurance/training/tasks/service/DeleteServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import com.codurance.training.tasks.Task;
+
+import java.io.PrintWriter;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+public class DeleteServiceImpl implements DeleteService{
+
+	private Map<String, List<Task>> tasks = new LinkedHashMap<>();
```
Comment (AbhilashZemoso) [inline java/src/main/java/com/codurance/training/tasks/service/DeleteServiceImpl.java]: make this final and remove initialisation

File: java/src/main/java/com/codurance/training/tasks/service/ProjectServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import com.codurance.training.tasks.Task;
+
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+public class ProjectServiceImpl implements ProjectService{
+
+	private Map<String, List<Task>> tasks = new LinkedHashMap<>();
```
Comment (AbhilashZemoso) [inline java/src/main/java/com/codurance/training/tasks/service/ProjectServiceImpl.java]: make this final and remove initialisation

File: java/src/main/java/com/codurance/training/tasks/service/TaskServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+
+public class TaskServiceImpl implements TaskService{
+
+	private Map<String, List<Task>> tasks = new LinkedHashMap<>();
```
Comment (AbhilashZemoso) [inline java/src/main/java/com/codurance/training/tasks/service/TaskServiceImpl.java]: make this final and remove initialisation

File: java/src/main/java/com/codurance/training/tasks/service/TaskServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+		return date;
+	}
+
+	private boolean checkIdValidity(String id) {
+		if(id.indexOf(" ") != -1)
+			return false;
+		Pattern pattern = Pattern.compile("[^a-zA-Z0-9]");
+		Matcher matcher = pattern.matcher(id);
+		return !(matcher.find());
+	}
```
Comment (AbhilashZemoso) [inline java/src/main/java/com/codurance/training/tasks/service/TaskServiceImpl.java]: create utility class and move private methods there

File: java/src/test/java/com/codurance/training/tasks/ApplicationTest.java [hunk 8]
```diff
@@ -15,113 +16,114 @@
...
+		}
+	}
+
+	private void write(String input) {
+		inWriter.println(input);
+	}
+
+	private boolean stillRunning() {
+		return applicationThread != null && applicationThread.isAlive();
+	}
```
Comment (AbhilashZemoso) [inline java/src/test/java/com/codurance/training/tasks/ApplicationTest.java]: Keep the existing test and add new tests on top of it
Don't update or remove the old ones

File: java/src/main/java/com/codurance/training/tasks/service/ShowServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+					out.printf("    [%c] %s: %s %s%n", (task.isDone() ? 'x' : ' '), task.getId(), task.getDescription(), parseDate(task.getDeadline()));
+			}
+			out.println();
+		}
+	}
+
+	private String parseDate(Date date) {
+		SimpleDateFormat formatter = new SimpleDateFormat("dd-MM-yyyy");
+		return formatter.format(date);
+	}
```
Comment (AbhilashZemoso) [inline java/src/main/java/com/codurance/training/tasks/service/ShowServiceImpl.java]: create utility class and move private method to that class

**Review by AbhilashZemoso — CHANGES_REQUESTED** [review]
work on comments

File: java/src/test/java/com/codurance/training/tasks/ApplicationTest.java (hunk 8 above)
Comment (AkashBoora) [reply java/src/test/java/com/codurance/training/tasks/ApplicationTest.java]: I had to change change abhilash, as we are passing idString when we add task

File: java/src/main/java/com/codurance/training/tasks/TaskList.java [hunk 10]
```diff
@@ -50,46 +54,12 @@ public void run() {
...
-                help();
-                break;
-            default:
-                error(command);
-                break;
-        }
-    }
 
-    private void add(String commandLine) {
+    public void add(String commandLine) {
```
Comment (AbhilashZemoso) [inline java/src/main/java/com/codurance/training/tasks/TaskList.java]: separate add related methods to a separate class

File: java/src/main/java/com/codurance/training/tasks/TaskList.java [hunk 11]
```diff
@@ -100,7 +70,7 @@ private void add(String commandLine) {
         }
     }
 
-    private void view(String command) {
+    public void view(String command) {
```
Comment (AbhilashZemoso) [inline java/src/main/java/com/codurance/training/tasks/TaskList.java]: move view, help, error method to a different Utility class

File: java/src/main/java/com/codurance/training/tasks/TaskList.java [hunk 12]
```diff
@@ -111,7 +81,7 @@ else if(command.equals("by project"))
             error(command);
     }
 
-    private void help() {
+    public void help() {
```
Comment (AbhilashZemoso) [inline java/src/main/java/com/codurance/training/tasks/TaskList.java]: move view, help, error method to a different Utility class

File: java/src/main/java/com/codurance/training/tasks/TaskList.java [hunk 13]
```diff
@@ -126,7 +96,7 @@ private void help() {
         out.println();
     }
 
-    private void error(String command) {
+    public void error(String command) {
```
Comment (AbhilashZemoso) [inline java/src/main/java/com/codurance/training/tasks/TaskList.java]: move view, help, error method to a different Utility class

File: java/src/main/java/com/codurance/training/tasks/service/Utility.java [hunk 14]
```diff
@@ -0,0 +1,11 @@
...
+
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+
+public interface Utility {
+	public Date parseDate(String deadline);
+	public String parseDateToString(Date date);
+	public boolean checkIdValidity(String id);
+}
```
Comment (AbhilashZemoso) [inline java/src/main/java/com/codurance/training/tasks/service/Utility.java]: Utility class should not have interfaces
Ref: http://blog.vanillajava.blog/2011/06/best-practice-dont-create-instances-of.html

File: java/src/main/java/com/codurance/training/tasks/service/UtilityImpl.java [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+		Pattern pattern = Pattern.compile("[^a-zA-Z0-9]");
+		Matcher matcher = pattern.matcher(id);
+		return !(matcher.find());
+	}
+
+	@Override
+	public String parseDateToString(Date date) {
+		SimpleDateFormat formatter = new SimpleDateFormat("dd-MM-yyyy");
+		return formatter.format(date);
+	}
```
Comment (AbhilashZemoso) [inline java/src/main/java/com/codurance/training/tasks/service/UtilityImpl.java]: make these public static

**Review by AbhilashZemoso — CHANGES_REQUESTED** [review]
work on comments

**Review by AbhilashZemoso — APPROVED** [review]
LGTM

**Pull request merged**